File: lokerpc/codegen/testdata/nullable.json.go
package typed

import (
	"context"

	"github.com/LOKE/pkg/lokerpc"
)

type NullableGadget *struct {
	ID string `json:"id"`
}

type Widget struct {
	ID string `json:"id"`
}

type GetNicknameResponse string

type GetTagsResponse []string

type GetUserRequest struct {
	ID string `json:"id"`
}

type GetUserResponse struct {
	ID string `json:"id"`
}

type SetNicknameRequest struct {
	ID string `json:"id"`
}

type SetTitleRequest string

type TypedService interface {
	GetGadget(context.Context, any) (NullableGadget, error)
	GetNickname(context.Context, any) (*GetNicknameResponse, error)
	GetTags(context.Context, any) (*GetTagsResponse, error)
	GetUser(context.Context, GetUserRequest) (*GetUserResponse, error)
	GetWidget(context.Context, any) (*Widget, error)
	SetGadget(context.Context, NullableGadget) error
	SetNickname(context.Context, *SetNicknameRequest) error
	SetTitle(context.Context, *SetTitleRequest) error
	SetWidget(context.Context, *Widget) error
}

type TypedRPCClient struct {
	lokerpc.Client
}

func (c TypedRPCClient) GetGadget(ctx context.Context, req any) (NullableGadget, error) {
	var res NullableGadget
	err := c.DoRequest(ctx, "getGadget", req, &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
func (c TypedRPCClient) GetNickname(ctx context.Context, req any) (*GetNicknameResponse, error) {
	var res *GetNicknameResponse
	err := c.DoRequest(ctx, "getNickname", req, &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
func (c TypedRPCClient) GetTags(ctx context.Context, req any) (*GetTagsResponse, error) {
	var res *GetTagsResponse
	err := c.DoRequest(ctx, "getTags", req, &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
func (c TypedRPCClient) GetUser(ctx context.Context, req GetUserRequest) (*GetUserResponse, error) {
	var res *GetUserResponse
	err := c.DoRequest(ctx, "getUser", req, &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
func (c TypedRPCClient) GetWidget(ctx context.Context, req any) (*Widget, error) {
	var res *Widget
	err := c.DoRequest(ctx, "getWidget", req, &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
func (c TypedRPCClient) SetGadget(ctx context.Context, req NullableGadget) error {
	return c.DoRequest(ctx, "setGadget", req, nil)
}
func (c TypedRPCClient) SetNickname(ctx context.Context, req *SetNicknameRequest) error {
	return c.DoRequest(ctx, "setNickname", req, nil)
}
func (c TypedRPCClient) SetTitle(ctx context.Context, req *SetTitleRequest) error {
	return c.DoRequest(ctx, "setTitle", req, nil)
}
func (c TypedRPCClient) SetWidget(ctx context.Context, req *Widget) error {
	return c.DoRequest(ctx, "setWidget", req, nil)
}

File: lokerpc/codegen/testdata/nullable.json.ts
import { RPCContextClient } from "@loke/http-rpc-client";
import { Context } from "@loke/context";

export type NullableGadget = {
  id: string;
} | null;

export type Widget = {
  id: string;
};

export type GetNicknameResponse = string | null;

export type GetTagsResponse = string[] | null;

export type GetUserRequest = {
  id: string;
};

export type GetUserResponse = {
  id: string;
} | null;

export type SetNicknameRequest = {
  id: string;
} | null;

export type SetTagsRequest = string[] | null;

export type SetTitleRequest = string | null;

/**
 * 
 */
export class TypedService extends RPCContextClient {
  constructor(baseUrl: string) {
    super(baseUrl, "typed")
  }
  /**
   * ref (not itself marked nullable) pointing at a definition that is nullable
   */
  getGadget(ctx: Context, req: any): Promise<NullableGadget> {
    return this.request(ctx, "getGadget", req);
  }
  /**
   * nullable scalar response
   */
  getNickname(ctx: Context, req: any): Promise<GetNicknameResponse> {
    return this.request(ctx, "getNickname", req);
  }
  /**
   * nullable array response
   */
  getTags(ctx: Context, req: any): Promise<GetTagsResponse> {
    return this.request(ctx, "getTags", req);
  }
  /**
   * inline nullable with properties
   */
  getUser(ctx: Context, req: GetUserRequest): Promise<GetUserResponse> {
    return this.request(ctx, "getUser", req);
  }
  /**
   * ref with nullable set on the ref usage itself, pointing at a non-nullable definition
   */
  getWidget(ctx: Context, req: any): Promise<Widget | null> {
    return this.request(ctx, "getWidget", req);
  }
  /**
   * ref (not itself marked nullable) pointing at a definition that is nullable (request side)
   */
  setGadget(ctx: Context, req: NullableGadget): Promise<void> {
    return this.request(ctx, "setGadget", req);
  }
  /**
   * nullable inline request
   */
  setNickname(ctx: Context, req: SetNicknameRequest): Promise<void> {
    return this.request(ctx, "setNickname", req);
  }
  /**
   * nullable array request
   */
  setTags(ctx: Context, req: SetTagsRequest): Promise<void> {
    return this.request(ctx, "setTags", req);
  }
  /**
   * nullable scalar request
   */
  setTitle(ctx: Context, req: SetTitleRequest): Promise<void> {
    return this.request(ctx, "setTitle", req);
  }
  /**
   * ref with nullable set on the ref usage itself, pointing at a non-nullable definition (request side)
   */
  setWidget(ctx: Context, req: Widget | null): Promise<void> {
    return this.request(ctx, "setWidget", req);
  }
}
